src/internal/determined/client.go

```diff
@@ -0,0 +1,140 @@
+	"github.com/determined-ai/determined/proto/pkg/userv1"
+	"github.com/determined-ai/determined/proto/pkg/workspacev1"
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	mlc "github.com/pachyderm/pachyderm/v2/src/internal/middleware/logging/client"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/metadata"
+)
+
+type Config struct {
+	MasterURL string
+	Username  string
+	Password  string
+	TLS       bool
+}
+
+func NewClient(ctx context.Context, masterURL string, withTLS bool) (client det.DeterminedClient, close func() error, retErr error) {
+	tlsOpt := grpc.WithTransportCredentials(insecure.NewCredentials())
+	if withTLS {
+		tlsOpt = grpc.WithTransportCredentials(credentials.NewTLS(&tls.Config{
+			InsecureSkipVerify: true,
+		}))
+	}
+	determinedURL, err := url.Parse(masterURL)
+	if err != nil {
+		return nil, nil, errors.Wrapf(err, "parsing determined url %q", masterURL)
+	}
+	conn, err := grpc.DialContext(ctx, determinedURL.Host, tlsOpt, grpc.WithStreamInterceptor(mlc.LogStream), grpc.WithUnaryInterceptor(mlc.LogUnary))
+	if err != nil {
+		return nil, nil, errors.Wrapf(err, "dialing determined at %q", determinedURL.Host)
+	}
+	dc := det.NewDeterminedClient(conn)
+	return dc, conn.Close, nil
+}
+
+func GetRole(ctx context.Context, dc det.DeterminedClient, name string) (*rbacv1.Role, error) {
+	resp, err := dc.ListRoles(ctx, &det.ListRolesRequest{})
+	if err != nil {
+		return nil, errors.Wrap(err, "search determined roles")
+	}
+	for _, r := range resp.Roles {
+		if r.Name == name {
+			return r, nil
+		}
+	}
+	return nil, errors.Errorf("workspaceEditor role not found")
+}
+
+func GetWorkspaces(ctx context.Context, dc det.DeterminedClient, workspaces []string) ([]*workspacev1.Workspace, error) {
+	workspacesResp, err := dc.GetWorkspaces(ctx, &det.GetWorkspacesRequest{})
+	if err != nil {
+		return nil, errors.Wrap(err, "list determined workspaces")
+	}
+	workspaceSet := make(map[string]struct{})
+	for _, w := range workspaces {
+		workspaceSet[w] = struct{}{}
+	}
+	res := make([]*workspacev1.Workspace, 0)
+	for _, dw := range workspacesResp.Workspaces {
+		if _, ok := workspaceSet[dw.Name]; ok {
+			res = append(res, dw)
+			delete(workspaceSet, dw.Name)
+		}
+	}
+	if len(workspaceSet) > 0 {
+		var errWorkspaces []string
+		for w := range workspaceSet {
+			errWorkspaces = append(errWorkspaces, w)
+		}
+		return nil, errors.Errorf("requested workspaces not found: %v", errWorkspaces)
+	}
+	return res, nil
+}
+
+func CreateWorkspace(ctx context.Context, dc det.DeterminedClient, workspace string) (*workspacev1.Workspace, error) {
+	w, err := dc.PostWorkspace(ctx, &det.PostWorkspaceRequest{Name: workspace})
+	if err != nil {
+		return nil, errors.Wrapf(err, "post determined workspace %q", workspace)
+	}
+	return w.Workspace, nil
+}
+
+// MintToken returns a context with a determined token loaded into its Metadata
```

Comment:
It _is_ efficient to have the flexibility of minting a token once and using it repeatedly.